lib/widgets/stream_colors.dart

```diff
@@ -0,0 +1,223 @@
+import 'dart:ui';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_color_models/flutter_color_models.dart';
+
+import '../api/model/model.dart';
+import 'color.dart';
+
+/// A provider for [StreamColorSwatch]es that computes and caches them.
+abstract class StreamColorSwatches {
+  static final StreamColorSwatches light = _StreamColorSwatchesLight._();
+  static final StreamColorSwatches dark = _StreamColorSwatchesDark._();
+
+  final Map<int, StreamColorSwatch> _cache = {};
+
+  /// Gives the cached [StreamColorSwatch] for a [subscription.color].
+  StreamColorSwatch forBaseColor(int base) =>
+    _cache[base] ??= computeForBaseColor(base);
+
+  StreamColorSwatch computeForBaseColor(int base);
+
+  /// Gives a [StreamColorSwatches], lerped to [other] at [t].
+  ///
+  /// If [this] and [other] are [identical], returns [this].
+  ///
+  /// Else returns an instance whose [forBaseColor] will call
+  /// [this.forBaseColor] and [other.forBaseColor]
+  /// and return [StreamColorSwatch.lerp]'s result on those.
+  /// This computation is cached on the instance
+  /// in order to save work building [t]'s animation frame when there are
+  /// multiple UI elements using the same [subscription.color].
+  StreamColorSwatches lerp(StreamColorSwatches? other, double t) {
```

Comment:
Perhaps make this static? That seems to be the convention elsewhere, or at least on `Color`.

(I had this thought when looking at the top of the class declaration, where there's `light` and `dark` and what one basically wants to say is that we have three kinds of instances of this class: light, dark, and lerps between them.)

Reply:
Sure, I can try that. I think I did non-static because that's what `ThemeExtension` does. (`CodeBlockTextStyles` came out non-static for that reason as well, I think.)

Reply:
`TextStyle`'s lerp is also static.

…Hmm, although I guess maybe `CodeBlockTextStyles` and `StreamColorSwatches` are more like `ThemeExtension` than like `Color` and `TextStyle`?

One thing about `CodeBlockTextStyles` and `StreamColorSwatches` is that their lerps are only ever called by a `ThemeExtension`'s lerp. In that lerp, the `a` instance (i.e., `this`) is non-nullable, and that's convenient for `_StreamColorSwatchesLerped.computeForBaseColor` because it makes it easy to reason locally that its non-nullable `StreamColorSwatch` return type is correct.

Reply:
> I think I did non-static because that's what `ThemeExtension` does.

Hmm I see.

I just did a quick survey of the Flutter tree. It looks like the pattern is very consistent, and `ThemeExtension` is an outlier. So let's follow the typical pattern and make it static.

Specifically the search I did was, in the Flutter repo:
```
$ git grep ' lerp[<(]' packages/flutter/lib/
```

That finds definitions of a method named `lerp`, static or otherwise.

There are a number of non-static `lerp` methods… on `Tween` subclasses and the like, that look like this:
```
packages/flutter/lib/src/rendering/tweens.dart
27:  FractionalOffset? lerp(double t) => FractionalOffset.lerp(begin, end, t);
50:  Alignment lerp(double t) => Alignment.lerp(begin, end, t)!;
75:  AlignmentGeometry? lerp(double t) => AlignmentGeometry.lerp(begin, end, t);
```
So e.g. `AlignmentTween.lerp` is non-static; but it's not a lerp between `AlignmentTween`s, rather a lerp between `Alignment`s, and in fact it forwards to the static `Alignment.lerp`.

There are something like 90 static `lerp` methods that take both endpoints as arguments, and like 40 non-static `lerp` methods that take neither as arguments, getting them both from the receiver (as with those Tween subclasses). And I think `ThemeExtension` is the sole example of its pattern, where it gets one endpoint from the receiver and the other as an argument.

Reply:
(This also means that that API choice on `ThemeExtension` was basically a mistake.)

Reply:
Thanks for that investigation!